renderer/backend/UniversalGpuBuffer.h

```diff
@@ -0,0 +1,55 @@
+/*****************************************************************************
+                    The Dark Mod GPL Source Code
+
+ This file is part of the The Dark Mod Source Code, originally based
+ on the Doom 3 GPL Source Code as published in 2011.
+
+ The Dark Mod Source Code is free software: you can redistribute it
+ and/or modify it under the terms of the GNU General Public License as
+ published by the Free Software Foundation, either version 3 of the License,
+ or (at your option) any later version. For details, see LICENSE.TXT.
+
+ Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+#pragma once
+
+class UniversalGpuBuffer {
+public:
+	void Init( GLenum type, GLuint size, GLuint alignment );
+	void Destroy();
+
+	void Lock();
+
+	byte *Reserve( GLuint size, bool precommit = false );
+	void Commit( byte *offset, GLuint size );
+	void BindRange( GLuint index, byte *offset, GLuint size );
+	void Bind();
+	const void * BufferOffset( const void *offset );
+
+private:
+	struct LockedRange {
+		GLuint offset;
+		GLuint count;
+		GLsync fenceSync;
```

Comment:
`Destroy` already clears any remaining fence syncs, not sure what the assert would accomplish?

Reply:
Not all structs are removed in `Destroy`, usually they are removed during normal operation. Right now one has to manually check that every such place deletes the fence. I imagine with a bit of changes those fences can start leaking, and that would be hard to notice. The assert would prevent that.

Reply:
Perhaps, but you'll never arrive in the destructor before first having called `Destroy` - at least not without yet another bug. So the assert in the destructor will not help. And you can't place the assert in `Destroy` because having some fences still pending is not an error.

Reply:
`WaitForLockedRange` destroys range directly. And it produces a feeling that it is perfectly OK to just write `mRangeLocks.erase(mRangeLocks.begin())` somewhere.

Reply:
I'm still not quite sure what it is you want me to do? Yes, the class is delicate because it manages GPU resources - if you alter it without knowing what you're doing you're going to cause issues. And I'm happy with whatever check catches potential errors, but the assert in the destructor will simply not do anything.

Only alternative I see is to abandon the concept of a circular buffer entirely and instead make it a "triple-range" buffer that uses one explicit range per frame. It would simplify the locking concept and abandon any issues with wraparound. On the downside it would mean that the buffer needs to be large enough to hold all data for the frame data of the worst-case scene with no wiggle room to deal with an occasional spike.

Reply:
As proposed, I've abandoned the circular buffer, so this circular locking concept is no longer needed.